Student Dashboard

mandryllo opened the pull request.

Commit:
Created student dashboard

Commit:
Created courses list view

Commit:
Created activity list view by type

Commit:
Cleanup :lipstick:

Commit:
Rendering teaching elements

Commit:
Navigation between TE siblings

Commit:
Cleanup and renaming

Commit:
Moved getSiblings to store getters

Commit:
Set active class on currect container

Commit:
CohortId in content api url is parameterized

Commit:
Course id is parameterized

Commit:
Fixed admin login :bug:

Commit:
Refactoring getter methods :wrench:

Commit:
Removed containers store module :goat:

Commit:
Cleanup :lipstick:

Commit:
Merge branch 'develop' into feature/student-dashboard

Commit:
Fixes and cleanup

underscope committed:
Merge branch 'develop' into feature/student-dashboard

Commit:
Created Student Root component

Commit:
Created Learner store module and learner home view

Commit:
Cleanup :lipstick:

Commit:
Changed learner home view design and more cleanup

Commit:
:goat:

Commit:
Separate learnerApi in learner and content APIs

Commit:
Cleanup :goat:

Commit:
Created truncate filter

Commit:
Select program is now only a mutation

Commit:
Created listUserEnrollments API endpoint

Commit:
Refactored courseware store getter :wrench:

underscope commented on client/student/api/contentApi.js:
```diff
@@ -0,0 +1,25 @@
+import request from '@/common/api/request';
+
+const extractData = res => res.data.data;
```
There is a helper for this within `/api/helpers`

underscope commented on client/student/api/contentApi.js:
```diff
@@ -0,0 +1,25 @@
+const url = {
+  syllabus: programId => `/cohorts/${programId}/content`,
+  container: (programId, courseId, containerId) => {
+    return `/cohorts/${programId}/content/${courseId}/container/${containerId}`;
```
Extract and combine base path. `courseId` should be renamed to `repositoryId`.

underscope commented on client/student/api/learnerApi.js:
```diff
@@ -0,0 +1,15 @@
+import request from '@/common/api/request';
+
+const extractData = res => res.data.data;
```
See prev comment regarding helper

underscope commented on client/student/api/learnerApi.js:
```diff
@@ -0,0 +1,15 @@
+const extractData = res => res.data.data;
+
+const url = {
+  programs: '/enrollments/currentUser'
```
Routes should be kebab-case

underscope commented on client/student/api/learnerApi.js:
```diff
@@ -0,0 +1,15 @@
+import request from '@/common/api/request';
```
Remove Api suffix from files within `/api`

underscope commented on client/student/components/Container/TeachingElements.vue:
```diff
@@ -0,0 +1,52 @@
+      :key="it.id"
+      :element="it"
+      :class="getElementWidth(it)"
+      class="column"/>
```
`.column` should be within `.columns` (parent; I saw you wrapped it parent component)

underscope commented on client/student/components/Container/TeachingElements.vue:
```diff
@@ -0,0 +1,52 @@
+  created() {
+    api.getContainer(this.selectedProgram, this.courseId, this.containerId)
+      .then(container => {
+        this.container = container;
```
Inline assignment in .then

underscope commented on client/student/components/Container/TeachingElements.vue:
```diff
@@ -0,0 +1,52 @@
+    activity() {
+      return find(this.courseware, it => it.container.id === this.containerId);
+    },
+    courseId() {
```
Seams like we are only using it within `created` hook so no need for computed.

underscope commented on client/student/components/Container/TeachingElements.vue:
```diff
@@ -0,0 +1,52 @@
+      return this.container.elements;
+    },
+    activity() {
+      return find(this.courseware, it => it.container.id === this.containerId);
```
Use shorthand `find(this.courseware, { container: { id: this.containerId } });`
This might be helpful on the store (currentActivity) as getter based on route params since other components might need this info.

underscope commented on client/student/components/Container/TeachingElements.vue:
```diff
@@ -0,0 +1,52 @@
+      return this.$route.params.containerId;
+    },
+    teachingElements() {
+      return this.container.elements;
```
Return blank array by default.

underscope commented on client/student/components/Container/TeachingElements.vue:
```diff
@@ -0,0 +1,52 @@
+  computed: {
+    ...mapState('learner', ['selectedProgram']),
+    ...mapGetters('learner', ['courseware']),
+    containerId() {
```
pass it as a prop

underscope commented on client/student/components/Container/TeachingElements.vue:
```diff
@@ -0,0 +1,52 @@
+export default {
+  name: 'teaching-elements',
+  data() {
+    return { container: {} };
```
Set to null by default.

underscope commented on client/student/components/Container/index.vue:
```diff
@@ -0,0 +1,58 @@
+<template>
```
Container is too generic name. Rename base to ContentContainer or some other better / more descriptive name.

underscope commented on client/student/components/Container/index.vue:
```diff
@@ -0,0 +1,58 @@
+<template>
+  <div class="container">
+    <nav class="columns navbar">
```
Make sure you are properly using columns

underscope commented on client/student/components/Container/index.vue:
```diff
@@ -0,0 +1,58 @@
+        v-for="it in siblings"
+        :key="it.id"
+        :to="{ name: 'container', params: { containerId: getId(it)}}"
+        :class="{ 'active-link': containerId === getId(it) }"
```
You can do this based on path. View router-link helpers (https://router.vuejs.org/api/#exact-active-class).

underscope commented on client/student/components/Container/index.vue:
```diff
@@ -0,0 +1,58 @@
+      </router-link>
+    </nav>
+    <div class="columns is-multiline">
+      <TeachingElements
```
component name should be kebab-case

underscope commented on client/student/components/Container/index.vue:
```diff
@@ -0,0 +1,58 @@
+      return this.$route.params.containerId;
+    },
+    activity() {
+      return find(this.courseware, it => it.container.id === this.containerId);
```
See my prev comment about this being needed multiple times (and available via store). In this case it might be also handled within parent (index) and passed as a prop (TeachingElements).

underscope commented on client/student/components/Container/index.vue:
```diff
@@ -0,0 +1,58 @@
+  name: 'container',
+  computed: {
+    ...mapGetters('learner', ['courseware']),
+    containerId() {
```
Pass it as a prop.

underscope commented on client/student/components/Container/index.vue:
```diff
@@ -0,0 +1,58 @@
+    activity() {
+      return find(this.courseware, it => it.container.id === this.containerId);
+    },
+    parentId() {
```
Not being reused and doesn't have any computation based on context. Please remove it as a computed.

underscope commented on client/student/components/Courseware/Card.vue:
```diff
@@ -0,0 +1,84 @@
+  },
+  methods: {
+    navigateTo() {
+      if (this.hasContainer) {
```
Why render it at all if it doesn't have content.

underscope commented on client/student/components/Courseware/index.vue:
```diff
@@ -0,0 +1,38 @@
+      <card
+        v-for="it in courseware"
+        :key="it._cid"
+        :courseware="it"/>
```
Bad name for prop since we are calling courseware whole content.

underscope commented on client/student/components/Courseware/index.vue:
```diff
@@ -0,0 +1,38 @@
+  },
+  methods: mapActions('learner', ['fetchSyllabus']),
+  created() {
+    this.fetchSyllabus(this.selectedProgram);
```
loader ?

underscope commented on server/enrollment/index.js:
```diff
@@ -4,6 +4,7 @@ const ctrl = require('./enrollment.controller');
 const router = require('express').Router();
 
 router
+  .get('/currentUser', ctrl.listUserEnrollments)
```
Routes should be kebab-case.

underscope commented on server/enrollment/enrollment.controller.js:
```diff
@@ -29,7 +29,14 @@ function create({ body }, res) {
     .then(enrollment => res.jsend.success(processOutput(enrollment)));
 }
 
+function listUserEnrollments({ user }, res) {
```
getUserEnrollments 💄

underscope commented on client/student/store/modules/learner/index.js:
```diff
@@ -0,0 +1,17 @@
+
+const state = {
+  programs: {},
+  selectedProgram: '',
```
Null

underscope commented on client/student/store/modules/learner/index.js:
```diff
@@ -0,0 +1,17 @@
+import * as mutations from './mutations';
+
+const state = {
+  programs: {},
```
Empty Array

underscope commented on client/student/store/modules/learner/index.js:
```diff
@@ -0,0 +1,17 @@
+const state = {
+  programs: {},
+  selectedProgram: '',
+  syllabus: {}
```
Empty array

underscope commented on client/student/store/modules/learner/getters.js:
```diff
@@ -0,0 +1,23 @@
+import forEach from 'lodash/forEach';
+import head from 'lodash/head';
+
+const filterBy = 'DEFAULT_SCHEMA/TOPIC';
```
CONST_CASE and rename so it reflects essence of configuration variable.

underscope commented on client/student/store/modules/learner/getters.js:
```diff
@@ -0,0 +1,23 @@
+const filterBy = 'DEFAULT_SCHEMA/TOPIC';
+
+export const courseware = state => {
+  let courseware = concat([], filter(state.syllabus, { schema: filterBy }));
```
You can omit concat

underscope commented on client/student/router.js:
```diff
@@ -1,13 +1,16 @@
+import Courseware from '@/student/components/Courseware/index';
 import ForgotPassword from '@/student/components/auth/ForgotPassword';
 import get from 'lodash/get';
+import LearnerHome from '@/student/components/Home/index';
```
This is within learner SPA, everything is Learner here (no need for prefix)

underscope commented on client/student/router.js:
```diff
@@ -34,10 +37,23 @@ const router = new Router({
-    component: StudentRoot,
-    meta: { auth: true }
+    path: '/',
+    name: 'learner-root',
```
See prev comment regarding naming....

underscope commented on client/student/components/common/Navbar.vue:
```diff
@@ -0,0 +1,32 @@
+
+<script>
+import { mapActions, mapState } from 'vuex';
+export default {
```
Missing some space

underscope commented on client/student/App.vue:
```diff
@@ -22,13 +22,17 @@ html, body {
+  display: flex;
   height: 100%;
+  padding: 32px 0;
+  flex-direction: column;
```
Do we need this within a root (flex / flex-column). Seams like it can be moved to the main container.

underscope commented on client/student/components/Home/index.vue:
```diff
@@ -0,0 +1,37 @@
+        <li
+          v-for="it in programs"
+          :key="it.id"
+          @click="selectProgram(it.cohortId)">
```
You can do it route based and selected program can be a getter (so you can omit this mutation). I know we initially talked about different approach but this might be nicer.

Commit:
Api cleanup

Commit:
Renamed components

Commit:
Container components cleanup

Commit:
Refactor courseware getter

Commit:
Fixed navigation on Home component and add loader on Courseware component

Commit:
Added Circular Progress component

Commit:
Refactored Learner Home component

Commit:
:lipstick:

Commit:
Merge branch 'develop' into feature/student-dashboard

Commit:
Fix import file after merge

Commit:
Renamed Card to ActivityCard

Commit:
Refactored courseware getter to return subActivities

Commit:
Moved container params to props

Commit:
Cleanup courseware getter

Commit:
Added dashboard button

Commit:
Cleanup

Commit:
Merge branch 'develop' into feature/student-dashboard

Commit:
Renamed TeachingElements component to Activity

Commit:
ContentContainer component cleanup

- renamed selectedProgram to selectedProgramId
- fixed CircularProgress style
- hide dashboard navbar link until program is selected

Commit:
Cleanup :lipstick:

- delete cohort controller
- renamed isSingleLevel getter to isCoursewareFlat
- refactor Activity component
- fixes CircularProgress styles
